### PR TITLE
[feature](executor) using fe version to set instance_num

### PR DESCRIPTION
**Pull request by Mryange**
## Proposed changes

Added files describing the process of FE upgrades.

'VersionFile' can be transformed like this.  
2_0_0-commitid-1-version -> 2_1_0-commitid-2-version ->2_3_0-commitid-3-version -> 2_0_0-commitid-4-version
You can observe the process of FE upgrades through these files.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.28 seconds
 stream load tsv:          507 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17162587840 Bytes

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java [hunk 1]
```diff
@@ -5329,6 +5338,45 @@ private static void addTableComment(Table table, StringBuilder sb) {
         }
     }
 
+    private void writeVersion() {
+        // "Write down the latest version of FE.
+        String fileName = this.metaDir + VERSION_FILE;
+        try (FileWriter fileWriter = new FileWriter(fileName)) {
+            fileWriter.write(CURRENT_FE_VERSION + "");
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java:5347]: use  logger.info("xxxxx",e)

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java [hunk 2]
```diff
@@ -289,6 +291,9 @@ public class Env {
     public static final String CLIENT_NODE_HOST_KEY = "CLIENT_NODE_HOST";
     public static final String CLIENT_NODE_PORT_KEY = "CLIENT_NODE_PORT";
 
+    private static final String VERSION_FILE = "/VERSION";
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java:294]: I think we could add a version_history folder in meta dir. And we could create multiple version files. like  1_2_3_commitid_version, 1_2_4_commitid_version,1_2_5_commitid_version, 1_2_6_commitid_version, 2_0_0_commitid_version
 
Then we could get the fe version history and commit id history.

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.6 seconds
 stream load tsv:          506 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17164090714 Bytes

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.43 seconds
 stream load tsv:          541 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17170265495 Bytes

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.95 seconds
 stream load tsv:          508 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17161969365 Bytes

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45 seconds
 stream load tsv:          534 seconds loaded 74807831229 Bytes, about 133 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          66 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17166650354 Bytes

**Comment by Mryange** [comment]
run feut

**Comment by Mryange** [comment]
run feut

**Comment by Mryange** [comment]
run feut

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java [hunk 3]
```diff
@@ -851,6 +859,7 @@ public void initialize(String[] args) throws Exception {
         this.metaDir = Config.meta_dir;
         this.bdbDir = this.metaDir + BDB_DIR;
         this.imageDir = this.metaDir + IMAGE_DIR;
+        this.versionDir = EnvUtils.getDorisHome() + VERSION_FILE;
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java:862]: VERSION_DIR

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java [hunk 4]
```diff
@@ -871,8 +880,23 @@ public void initialize(String[] args) throws Exception {
             }
         }
         File imageDir = new File(this.imageDir);
+        File verDir = new File(this.versionDir);
+
         if (!imageDir.exists()) {
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java:885]: if (!imageDir.exists) {
imageDir.mkdirs();
}
if (!versionDir.exists()) {
versionDir.mkdirs();
}

getPreviousVersion(); // previous version == unknown;
{
 files[] aa = listdir();
  aa. name .split('-')[3]
  
previousVersion = xxx;

}
writeLatestVersion() {
// if currentVersion == previousVersion
else {}

}

1_2_0-commitid-seq-version

1.2.0-->1.2.1-->1.2.0

**Comment by Mryange** [comment]
run feut

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.3 seconds
 stream load tsv:          513 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17166321231 Bytes

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.89 seconds
 stream load tsv:          540 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17168137204 Bytes

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java [hunk 5]
```diff
@@ -5330,6 +5344,58 @@ private static void addTableComment(Table table, StringBuilder sb) {
...
+        return previousFeVersion.charAt(0) != latestFeVersion.charAt(0);
+    }
+
+    private void initVersionInfo() {
+        File folder = new File(versionDir);
+        File[] files = folder.listFiles();
+        int previousSeq = 0;
+        if (files != null) {
+            for (File file : files) {
+                String[] splitArr = file.getName().split("-");
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java:5378]: add comment here to describe the version file name and every part meaning

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java [hunk 6]
```diff
@@ -290,10 +291,13 @@ public class Env {
     public static final String CLIENT_NODE_HOST_KEY = "CLIENT_NODE_HOST";
     public static final String CLIENT_NODE_PORT_KEY = "CLIENT_NODE_PORT";
 
+    private static final String VERSION_DIR = "/VERSION";
+    public static String latestFeVersion = "2_0_0";
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java:295]: do not add this variable, should generate the version from global variables. For example, show variable will show version info.
You add this variable and then when we publish a version, we has to remember to modify this variable. It is very difficult to maintain it.

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.53 seconds
 stream load tsv:          506 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17161956801 Bytes

File: fe/fe-core/src/main/java/org/apache/doris/qe/ConnectContext.java [hunk 7]
```diff
@@ -263,6 +263,9 @@ public ConnectContext(StreamConnection connection) {
             mysqlChannel = new DummyMysqlChannel();
         }
         sessionVariable = VariableMgr.newSessionVariable();
+        if (isMajorVersionUpgrade) {
```
Comment (HappenLee) [inline fe/fe-core/src/main/java/org/apache/doris/qe/ConnectContext.java:266]: we need global set, here just change the connect way, if fe restart will change the logic

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.94 seconds
 stream load tsv:          507 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162553758 Bytes

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.68 seconds
 stream load tsv:          508 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162057178 Bytes

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by zhangstar333 — APPROVED** [review]

**Pull request merged**